aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,329 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes;
+    tensor_view_sizes.push_back(-1);
+    for (int d = n_batch_dims; d < tensor.dim(); ++d) {
+      tensor_view_sizes.push_back(tensor.size(d));
+    }
+    auto tensor_view = tensor.view(tensor_view_sizes);
+    tensors_view.push_back(tensor_view);
+  }
+
+  for (int64_t input_idx = 0; input_idx < tensors_view[0].size(0);
+    ++input_idx) {
+    std::vector<Tensor> input_selected;
+    for (auto& tensor_view : tensors_view) {
+      auto curr_input = tensor_view.select(0, input_idx);
+      input_selected.push_back(curr_input);
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
+  if (matrices.dim() > 2) {
+    auto res = at::empty(matrices.sizes(), matrices.options());
+
+    batch_apply(
+      {res, matrices, powers},
+      matrices.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+        int64_t n = tensors[2].item().template to<int64_t>();
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + A2 / 2;
+}
+
+Tensor compute_T4(const Tensor& A) {
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / 2. + A / 6. + A2 / 24.);
+}
+
+template <typename scalar_t>
+Tensor compute_T8(const Tensor& A) {
+  // no constexpr support for std::sqrt unfortunately
+  scalar_t sqrt_177 = std::sqrt(177);
+  scalar_t x3 = 2. / 3.;
+  scalar_t x1 = x3 * ((1 + sqrt_177) / 88);
+  scalar_t x2 = x3 * ((1 + sqrt_177) / 352);
+  scalar_t x4 = (-271 + 29 * sqrt_177) / (315 * x3);
+  scalar_t x5 = (-11 + 11 * sqrt_177) / (1260 * x3);
+  scalar_t x6 = (-99 + 11 * sqrt_177) / (5040 * x3);
+  scalar_t x7 = (89 - sqrt_177) / (5040 * x3);
+  scalar_t y2 = (857 - 58 * sqrt_177) / 630;
+
+  auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  auto A2 = at::matmul(A, A);
+  auto A4 = at::matmul(A2, x1 * A + x2 * A2);
+  auto A8 = at::matmul(x3 * A2 + A4, x4 * I + x5 * A + x6 * A2 + x7 * A4);
+  return I + A + y2 * A2 + A8;
+}
+
+template <typename scalar_t>
+Tensor compute_T12(const Tensor& A) {
+  constexpr scalar_t b[][4] = {
+    {
+      9.0198e-16,
```

Comment:
This is very weird. These are the coefficients in the paper, but they are not the same as the coefficients they used in their matlab implementation... I've just sent an email to the authors of the paper to see if they can clarify where's this mismatch coming from...

Reply:
I see, I saw that difference in your code versus their paper. So, you used their matlab code as a reference?

Reply:
Yes, I took their matlab code and ported it to PyTorch, as I assumed that they had run all their tests with that code.

Reply:
@Lezcano, is there any update from the authors?

Reply:
Nothing yet... Have you tried comparing against T12 computed as I + A + A^2/2 + ... to see if one of them is wrong?

Reply:
The tests do that already, and it works just fine

Reply:
But, anyway, the get those coefficients by solving a system of non-linear equations, right? Well, I am sure then uniqueness is not assumed

Reply:
The tests do not test explicitly each approximation implemented, right? In most of the cases, the approximation used is just T18. It happened to me when I implemented this that I had a small mismatch in a coefficient in T8 and I didn't detect that until I tested each approximation individually.

Reply:
Nevermind, I just saw that you do test that. It might be what you said about the uniqueness, yep.

Reply:
The tests do check it, see the renormalization loop